### PR TITLE
Execute gas msg

### LINKED ISSUE #579: Allow data request to specify the amount of gas allocated for `execute`
Note that this diff does not involve changing end block logic.

### PR DESCRIPTION
**Pull request by taobun**
fixed #579

### CODE REVIEW

File: chain/cmd/bandsv/main.go [hunk 1]
```diff
@@ -32,6 +32,8 @@ const (
 	DefaultRequestedValidatorCount  = 1
 	DefaultSufficientValidatorCount = 1
 	DefaultExpiration               = 100
+
+	DefaultExecuteGas = 50000
```
Comment (sorawit) [inline chain/cmd/bandsv/main.go]: Why have extra space?

File: chain/x/zoracle/internal/keeper/request.go [hunk 2]
```diff
@@ -121,6 +121,21 @@ func (k Keeper) CheckRequestExists(ctx sdk.Context, id int64) bool {
 	return store.Has(types.RequestStoreKey(id))
 }
 
+// ConsumeGasForExecute saves and consumes gas from the current context to use in executing
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/request.go]: Good job. Just small nit: use for the execution at ...

File: chain/x/zoracle/internal/keeper/request.go [hunk 3]
```diff
@@ -121,6 +121,21 @@ func (k Keeper) CheckRequestExists(ctx sdk.Context, id int64) bool {
...
+// at end block time.
+func (k Keeper) ConsumeGasForExecute(ctx sdk.Context, id int64, executeGas uint64) sdk.Error {
+	request, err := k.GetRequest(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	ctx.GasMeter().ConsumeGas(executeGas, "Execute gas")
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/request.go]: I think the convention is to do dash? "execute-gas"

File: chain/x/zoracle/handler.go [hunk 4]
```diff
@@ -155,6 +155,11 @@ func handleMsgRequestData(ctx sdk.Context, keeper Keeper, msg MsgRequestData) sd
 		return err.Result()
 	}
 
+	err = keeper.ConsumeGasForExecute(ctx, id, msg.ExecuteGas)
```
Comment (sorawit) [inline chain/x/zoracle/handler.go]: IMO the consume gas can just happen in `AddRequest`. By doing that also avoid unnecessary load-and-update of `Request`.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/keeper/request.go (hunk 3 above)
Comment (taobun) [reply chain/x/zoracle/internal/keeper/request.go]: Change to "ExecuteGas"

**Review by sorawit — DISMISSED** [review]

File: chain/x/zoracle/client/cli/tx.go [hunk 5]
```diff
@@ -120,6 +126,8 @@ $ %s tx zoracle request 1 --calldata 1234abcdef --requested-validator-count 4 --
 	cmd.MarkFlagRequired(flagSufficientValidatorCount)
 	cmd.Flags().Int64P(flagExpiration, "x", 0, "Maximum block count before the data request is considered expired")
 	cmd.MarkFlagRequired(flagExpiration)
+	cmd.Flags().Uint64P(flagExecuteGas, "g", 0, "The amount of gas will be consumed for later execution")
```
Comment (sorawit) [inline chain/x/zoracle/client/cli/tx.go]: that will be reserved

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/client/cli/tx.go [hunk 6]
```diff
@@ -53,7 +54,7 @@ func GetTxCmd(storeKey string, cdc *codec.Codec) *cobra.Command {
 // GetCmdRequest implements the request command handler.
 func GetCmdRequest(cdc *codec.Codec) *cobra.Command {
 	cmd := &cobra.Command{
-		Use:   "request [oracle-script-id] (-c [calldata]) (-r [requested-validator-count]) (-v [sufficient-validator-count]) (-x [expiration])",
+		Use:   "request [oracle-script-id] (-c [calldata]) (-r [requested-validator-count]) (-v [sufficient-validator-count]) (-x [expiration]) (-g [execute gas])",
```
Comment (evilpeach) [inline chain/x/zoracle/client/cli/tx.go]: - edit `execute gas` to `execute-gas`

File: chain/x/zoracle/client/cli/tx.go [hunk 7]
```diff
@@ -120,6 +126,8 @@ $ %s tx zoracle request 1 --calldata 1234abcdef --requested-validator-count 4 --
 	cmd.MarkFlagRequired(flagSufficientValidatorCount)
 	cmd.Flags().Int64P(flagExpiration, "x", 0, "Maximum block count before the data request is considered expired")
 	cmd.MarkFlagRequired(flagExpiration)
+	cmd.Flags().Uint64P(flagExecuteGas, "g", 0, "The amount of gas that will be reserved for later execution")
```
Comment (evilpeach) [inline chain/x/zoracle/client/cli/tx.go]: change `g` to `eg`

**Review by evilpeach — CHANGES_REQUESTED** [review]

**Review by sorawit — DISMISSED** [review]

**Review by evilpeach — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**